libbpf-tools/bashreadline.c

```diff
@@ -0,0 +1,232 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+#include <argp.h>
+#include <stdio.h>
+#include <errno.h>
+#include <signal.h>
+#include <time.h>
+#include <unistd.h>
+#include <ctype.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "bashreadline.h"
+#include "bashreadline.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+
+static volatile sig_atomic_t exiting = 0;
+
+const char *argp_program_version = "bashredline 1.0";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Print entered bash commands from all running shells.\n"
+"\n"
+"USAGE: bashreadline [-s <readline>]\n"
+"\n"
+"EXAMPLES:\n"
+"    bashreadline\n"
+"    bashreadline -s /usr/lib/libreadline.so\n";
+
+static const struct argp_option opts[] = {
+	{ "shared", 's', "PATH", 0, "the location of libreadline.so library" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static char *libreadline_path = NULL;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 's':
+		libreadline_path = strdup(arg);
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_size)
+{
+	struct str_t *e = data;
+	struct tm *tm;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%m:%S", tm);
+
+	printf("%-9s %-6lld %s\n", ts, e->pid, e->str);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static char *find_readline_so() {
+	const char *bash_path = "/bin/bash";
+	FILE *fp;
+	char *line = NULL;
+	char *sep, *path_start, *sp;
+	size_t line_sz;
+	off_t func_off;
+	char *result = NULL;
+
+	func_off = get_elf_func_offset(bash_path, "readline");
+	if (func_off >= 0) {
+		return strdup(bash_path);
+	}
+
+	/*
+	 * Try to find libreadline.so if the symbol of readline is not
+	 * defined in bash
+	 */
+	fp = popen("ldd /bin/bash", "r");
+	if (fp == NULL)
+		goto cleanup;
+	while (getline(&line, &line_sz, fp) >= 0) {
+		if (strstr(line, "libreadline.so") == NULL)
+			continue;
+		sep = strstr(line, " => ");
```

Comment:
Use `sscanf` ?

Reply:
Although this is not a perfect parser, sscanf() is not adaptive for changes in the output format.  For example, the length of leading spaces may change.  Does that make sense?

Reply:
It's OK. I don't have strong opinion on this.

Reply:
ok, so now that I see what we are parsing, I think sscanf actually makes total sense. You were worried about sscanf() not handling leading spaces. That's not true. If you specify format string as "%s => %s" sscanf() will skip any number of whitespaces (tabs and spaces, maybe some other whitespaces) between the first string, "=>", and second string. So I think sscanf("") makes sense here. It will make code much shorter and more understandable.